lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -543,8 +593,74 @@ int MySQL_HostGroups_Manager::get_multiple_idle_connections(int _hid, unsigned l
 		}
 	}
 __exit_get_multiple_idle_connections:
+	status.myconnpoll_get_ping+=num_conn_current;
 	wrunlock();
 	proxy_debug(PROXY_DEBUG_MYSQL_CONNPOOL, 7, "Returning %d idle connections\n", num_conn_current);
 	return num_conn_current;
 }
 
+SQLite3_result * MySQL_HostGroups_Manager::SQL3_Connection_Pool() {
+  const int colnum=6;
+  proxy_debug(PROXY_DEBUG_MYSQL_CONNECTION, 4, "Dumping Connection Pool\n");
+  SQLite3_result *result=new SQLite3_result(colnum);
+  result->add_column_definition(SQLITE_TEXT,"hostgroup");
+  result->add_column_definition(SQLITE_TEXT,"srv_host");
+  result->add_column_definition(SQLITE_TEXT,"srv_port");
+  result->add_column_definition(SQLITE_TEXT,"status");
+  result->add_column_definition(SQLITE_TEXT,"ConnUsed");
+  result->add_column_definition(SQLITE_TEXT,"ConnFree");
+
+	wrlock();
+	int i,j, k;
+	for (i=0; i<(int)MyHostGroups->len; i++) {
+		MyHGC *myhgc=(MyHGC *)MyHostGroups->index(i);
+		for (j=0; j<(int)myhgc->mysrvs->cnt(); j++) {
+			MySrvC *mysrvc=(MySrvC *)myhgc->mysrvs->servers->index(j);
+			if (mysrvc->status!=MYSQL_SERVER_STATUS_ONLINE) {
+				proxy_debug(PROXY_DEBUG_MYSQL_CONNPOOL, 5, "Server %s:%d is not online\n", mysrvc->address, mysrvc->port);
+				mysrvc->ConnectionsFree->drop_all_connections();
+			}
+			// drop idle connections if beyond max_connection
+			while (mysrvc->ConnectionsFree->conns->len && mysrvc->ConnectionsUsed->conns->len+mysrvc->ConnectionsFree->conns->len > mysrvc->max_connections) {
```

Comment:
It's not intuitive at all to do this kind of clean-up in here. This is only supposed to introspect what is going on inside the data structures, not also modify them. "Principle of least surprise"